### PR TITLE
chore(android): improve gradle build fault tolerance

### PR DESCRIPTION
**Pull request by jquick-axway**
**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-27718
- https://jira.appcelerator.org/browse/TIMOB-27720

**Summary:**
- [TIMOB-27718](https://jira.appcelerator.org/browse/TIMOB-27718) Log build warnings if "res" files have invalid names.
  * Restores old Titanium behavior where "res" file names are automatically renamed if invalid.
  * Will log build warnings if invalid along with a suggestion on what to rename it to.
- [TIMOB-27720](https://jira.appcelerator.org/browse/TIMOB-27720) Titanium SDK build should favor "python2" tool if available.
  * Avoids SDK build failures on systems that have python 3 installed.
  * Note: We plan on removing our python usage in the future.
- Now setting Java/Kotlin compiler max heap size via gradle.
  * Avoids `OutOfMemory` exception when building very large projects.
  * Uses existing `android.javac.maxmemory` setting.
  * Uses existing `android.dx.maxmemory` setting.
- Improved "AndroidManifest.xml" fault tolerance for app builds.
  * Now filters out `xmlns:android` namespaces attributes from child XML elements. Can only be defined in root `<manifest/>` element. CLI used to wrongly inject them.

---
**Test:**
1. Create a Classic Titanium app project.
2. Unzip "Resources.zip" attached to [TIMOB-27718](https://jira.appcelerator.org/browse/TIMOB-27718) to project's "Resources" folder.
3. Copy below "tiapp.xml" settings to project.
4. Copy below "app.js" to project.
5. Build and run on Android.
6. Verify that you see 2 Appc logos and a gray button image with dark border onscreen.
7. Verify that you see the following build warnings in the log.
```
[WARN]  Invalid "res" file: Resources/android/images/res-mdpi/1-appc-dev-logo.jpg
[WARN]  - Names must contain only lowercase a-z, 0-9, or underscore.
[WARN]  - Names cannot start with a number.
[WARN]  - Titanium will rename to: _1_appc_dev_logo.jpg
[WARN]  Invalid "res" file: Resources/android/images/res-mdpi/ButtonHoloDark.9.png
[WARN]  - Names must contain only lowercase a-z, 0-9, or underscore.
[WARN]  - Titanium will rename to: buttonholodark.9.png
[WARN]  Invalid "res" file: Resources/android/images/res-mdpi/Logos/Appc-Transparent.png
[WARN]  - Files cannot be put into subdirectories.
[WARN]  - Names must contain only lowercase a-z, 0-9, or underscore.
[WARN]  - Titanium will rename to: logos_appc_transparent.png
```

tiapp.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ti:app xmlns:ti="http://ti.appcelerator.org">
	<android xmlns:android="http://schemas.android.com/apk/res/android">
		<manifest>
			<uses-permission xmlns:android="" android:name="android.permission.FOREGROUND_SERVICE"/>
			<uses-permission xmlns:android="http://schemas.android.com/apk/res/android" android:name="android.permission.WAKE_LOCK"/>
			<application xmlns:android="">
				<meta-data xmlns:android="" android:name="my.metadata" android:value="Hello World"/>
			</application>
		</manifest>
	</android>
</ti:app>
```

app.js
```javascript
var imagePath;
var window = Ti.UI.createWindow();
var scrollView = Ti.UI.createScrollView({
	layout: "vertical",
	scrollType: "vertical",
	showHorizontalScrollIndicator: false,
	shorVerticalScrollIndicator: true,
	backgroundColor: "gray",
});
{
	imagePath = "/Resources/images/1-appc-dev-logo.jpg";
	scrollView.add(Ti.UI.createLabel({ text: imagePath }));
	scrollView.add(Ti.UI.createImageView({ image: imagePath }));
}
{
	imagePath = "/Resources/images/Logos/Appc-Transparent.png";
	scrollView.add(Ti.UI.createLabel({ text: imagePath }));
	scrollView.add(Ti.UI.createImageView({ image: imagePath }));
}
{
	imagePath = "/Resources/images/ButtonHoloDark.9.png";
	scrollView.add(Ti.UI.createLabel({ text: imagePath }));
	scrollView.add(Ti.UI.createView({ backgroundImage: imagePath, width: "50%", height: "25%" }));
}
window.add(scrollView);
window.open();
```

### CODE REVIEW

File: android/cli/lib/android-manifest.js [hunk 1]
```diff
@@ -598,20 +598,30 @@ class AndroidManifest {
...
+			attributeIndex = text.indexOf('xmlns:android', manifestIndex);
+			if ((attributeIndex < 0) || (attributeIndex > text.indexOf('>', manifestIndex))) {
 				const attributeString = ' xmlns:android="http://schemas.android.com/apk/res/android"';
 				text = text.substring(0, insertIndex) + attributeString + text.substring(insertIndex);
 			}
+
+			// Remove "xmlns:android" namespace attributes from all child elements. Only supported in <manifest/>.
+			// Note: CLI used to wrongly inject these after updating "tiapp.xml" if missing from root element.
+			const manifestEndIndex = text.indexOf('>', manifestIndex);
+			text = text.replace(/xmlns:android=".*?"/g, (match, p1) => {
```
Comment (garymathews) [inline android/cli/lib/android-manifest.js]: In the regex `/xmlns:android=".*?"/g` is the `?` necessary, wouldn't the `.*` suffice?

File: android/cli/commands/_build.js [hunk 2]
```diff
@@ -898,12 +898,9 @@ AndroidBuilder.prototype.validate = function validate(logger, config, cli) {
 	}
 	cli.tiapp.properties['ti.deploytype'] = { type: 'string', value: this.deployType };
 
-	// Titanium's gradle build system currently ignores these settings.
+	// Fetch Java max heap size settings.
 	this.javacMaxMemory = cli.tiapp.properties['android.javac.maxmemory'] && cli.tiapp.properties['android.javac.maxmemory'].value || config.get('android.javac.maxMemory', '3072M');
-	this.javacSource = cli.tiapp.properties['android.javac.source'] && cli.tiapp.properties['android.javac.source'].value || config.get('android.javac.source', '1.7');
-	this.javacTarget = cli.tiapp.properties['android.javac.target'] && cli.tiapp.properties['android.javac.target'].value || config.get('android.javac.target', '1.7');
 	this.dxMaxMemory = cli.tiapp.properties['android.dx.maxmemory'] && cli.tiapp.properties['android.dx.maxmemory'].value || config.get('android.dx.maxMemory', '3072M');
-	this.dxMaxIdxNumber = cli.tiapp.properties['android.dx.maxIdxNumber'] && cli.tiapp.properties['android.dx.maxIdxNumber'].value || config.get('android.dx.maxIdxNumber', '65536');
```
Comment (garymathews) [inline android/cli/commands/_build.js]: Do we want to remove the ability to set `maxIdxNumber`?

File: android/cli/lib/android-manifest.js (hunk 1 above)
Comment (jquick-axway) [reply android/cli/lib/android-manifest.js]: Yes, the `?` is needed. Without it, the regex goes past the attribute and to the last double quote before the next match.

File: android/cli/commands/_build.js (hunk 2 above)
Comment (jquick-axway) [reply android/cli/commands/_build.js]: The `d8` command line tool doesn't appear to support this. I ran `--help` on it and didn't find an equivalent.

**Review by garymathews — APPROVED** [review]
CR: **PASS**

**Review by ssjsamir — APPROVED** [review]
FR Passed: Tested using the test steps mentioned in the description above. 

**Test Environment**
```
MacOS Catalina: 10.15.1 beta
Xcode: 11.3
Java Version: 1.8.0_131
Android NDK: 18.1.5063045
Node.js: 10.16.3
"NPM":"4.2.15-1","CLI":"7.1.2-7"
Pixel xl (7.1.1) Sim
```

**Pull request merged**